types/types.go

```diff
@@ -170,6 +170,13 @@ type BICReplacementCandidate struct {
 	Location BICLocationReference
 }
 
+// BICReplacementCandidate2 is an item returned by BlobInfoCache2.CandidateLocations2.
+type BICReplacementCandidate2 struct {
+	Digest         digest.Digest
+	CompressorName string
```

Comment:
Can this be “uncompressed” or “unknown”? If so, that should be documented and distinct.

And do we just soldier on when getting “unknown”, or prefer correctness and don’t reuse? Just by asking that question I’m somewhat suggesting to prefer correctness, OTOH that would make third-party cache implementations half-ignored. Something for me to think about while I continue reading tomorrow.

Reply:
Yes, I think it makes sense to mark cases where we don't know because we retrieved the data from a cache that original methods.  Added constants for uncompressed and unknown to avoid ambiguity.

Reply:
Distinguishing cases where the compression type isn't known from uncompressed cases would keep us from actively breaking things, so I'm leaning that way as well.  Updated `dockerImageDestination.TryReusingBlob()` to skip candidate locations where the compression type isn't known.

Reply:
Okay, I _think_ we don't need to skip these any more.

Reply:
What happens with unknown compression now? If that happens (i.e. an old BoltDB cache that knows a location but no compression, or a third-party `BlobInfoCache` implementation), don’t we fall back to the old behavior of using that blob but not updating the MIME type to match?

Reply:
`OperationAndAlgorithmForCompressor()` gets passed `UnknownCompression` and returns `PreserveOriginal` with no specified compressor, and `dockerImageDestination.TryReusingBlob()` returns them in a `BlobInfo` with the original media type and no compression changes.  When that eventually gets to `manifest.updatedMIMEType()`, it doesn't make any changes.  So until we actually read that blob again, we will continue to behave as before when reusing that blob.

Reply:
Isn’t “operate as before” #733? Some version of this PR skipped the blobs with `UnknownCompression`: https://github.com/containers/image/pull/1089#discussion_r539574676 .

That might force an extra unnecessary upload for schema1 (where the MIME type does not matter), but with schema1 disabled by docker/distribution by default, fixing #733 more reliably seems preferable.